hudi-metaserver/src/main/java/org/apache/hudi/metaserver/client/HoodieMetaServerClientImp.java

```diff
@@ -0,0 +1,194 @@
+    for (int i = 0; !isConnected && i < retryLimit; i++) {
+      try {
+        URI msUri = new URI(uri);
+        this.transport = new TSocket(msUri.getHost(), msUri.getPort());
+        this.client = new ThriftHoodieMetaServer.Client(new TBinaryProtocol(transport));
+        transport.open();
+        this.isConnected = true;
+        LOG.info("Connected to meta server: " + msUri);
+      } catch (URISyntaxException e) {
+        throw new HoodieException("Invalid meta server uri: " + uri, e);
+      } catch (TTransportException e) {
+        exception = e;
+        LOG.warn("Fail to connect to the meta server.", e);
+      }
+    }
+    if (!isConnected) {
+      throw new HoodieException("Fail to connect to the meta server.", exception);
+    }
+  }
+
+  private boolean isLocalEmbeddedMetaServer(String uri) {
+    return uri == null ? true : uri.trim().isEmpty();
+  }
+
+  @Override
+  public Table getTable(String db, String tb) {
+    return exceptionWrapper(() -> this.client.get_table(db, tb)).get();
+  }
+
+  @Override
+  public void createTable(Table table) {
+    try {
+      this.client.create_table(table);
+    } catch (TException e) {
+      throw new HoodieException(e);
+    }
+  }
+
+  public List<HoodieInstant> listInstants(String db, String tb, int commitNum) {
+    return exceptionWrapper(() -> this.client.list_instants(db, tb, commitNum).stream()
+        .map(EntryConvertor::fromTHoodieInstant)
+        .collect(Collectors.toList())).get();
+  }
+
+  public Option<byte[]> getInstantMeta(String db, String tb, HoodieInstant instant) {
+    ByteBuffer bytes = exceptionWrapper(() -> this.client.get_instant_meta(db, tb, EntryConvertor.toTHoodieInstant(instant))).get();
+    Option<byte[]> res = bytes.capacity() == 0 ? Option.empty() : Option.of(bytes.array());
+    return res;
+  }
+
+  public String createNewTimestamp(String db, String tb) {
+    return exceptionWrapper(() -> this.client.create_new_instant_time(db, tb)).get();
+  }
+
+  public void createNewInstant(String db, String tb, HoodieInstant instant, Option<byte[]> content) {
+    exceptionWrapper(() -> this.client.create_new_instant_with_time(db, tb, EntryConvertor.toTHoodieInstant(instant), getByteBuffer(content))).get();
+  }
+
+  public void transitionInstantState(String db, String tb, HoodieInstant fromInstant, HoodieInstant toInstant, Option<byte[]> content) {
+    exceptionWrapper(() -> this.client.transition_instant_state(db, tb,
+        EntryConvertor.toTHoodieInstant(fromInstant),
+        EntryConvertor.toTHoodieInstant(toInstant),
+        getByteBuffer(content))).get();
+  }
+
+  public void deleteInstant(String db, String tb, HoodieInstant instant) {
+    exceptionWrapper(() -> this.client.delete_instant(db, tb, EntryConvertor.toTHoodieInstant(instant))).get();
+  }
+
+  // TODO: support snapshot creation
+  public FileStatus[] listFilesInPartition(String db, String tb, String partition, String timestamp) {
+    return null;
```

Comment:
Better return empty array instead of null

Reply:
It's just a reserved api here for supporting the snapshot and filesystem creation. will fix in the next following pr.